### PR TITLE
Documentation in printable format (CSS media queries)

### PR DESCRIPTION
**Pull request by SetTrend**
### Synopsis
With this PR implemented, documentation becomes printable; i.e. without navigation, comment and other chrome.

### Steps done
1. New SASS rules:
   1. All `.interactive` sections get hidden for printing.
   1. The `contributors` section gets hidden for printing.
   1. The `.footer` gets hidden for printing.
   1. The `.page-links` section gets hidden for printing.
1. I created a `NodeJS` script that recursively concatenates all MarkDown files within a folder into large "chapter" files. The new, automatically generated files match the regular expression: "`(_\w+)?_all.md`".

### Further Information
At this time I have no clue on how to present the generated "`*_all.md`" files to the user. Perhaps you've got an idea?

### CODE REVIEW

**Comment by SetTrend** [comment]
@montogeek : Want to approve?

**Comment by EugeneHlushko** [comment]
Hi @SetTrend, we probably want to get `rebuild` into master first, let me catch up with @montogeek on this.

**Comment by EugeneHlushko** [comment]
Also i think we dont want to commit the aggregated files, rather build them before deploy

**Comment by SetTrend** [comment]
### The aggregated files are important for printing, for the following reasons:

- I didn't add a TOC yet, but adding a TOC wouldn't make sense for 2-3 pages, each.

- Printing commonly is done in Brochure format, i.e. 2 sides per page:
  brochure printing - Adobe Acrobat Reader

  Printing each of the documentation's web pages separately would leave too many pages unnecessarily blank:
  brochure printing, single page - Adober Acrobat Reader
<br/>

### I suggest to add a "Print as PDF" link to the documentation navigation pages
... which in turn would open the concatenated files and immediately open a printing dialog, i.e. calling `window.print();`.

Microsoft ASP.NET does the same:

asp net core - documentation

**Comment by EugeneHlushko** [comment]
Yes files are important, but they can be generated on the fly before deployment from the fresh revision. We dont need to commit and ignore them

**Comment by SetTrend** [comment]
Ah, sure, I now understand. So, you just don't want to have the "artifacts" stored in the repository. That's a great idea.

I can easily remove the files from my branch. However, I'm not too familiar with the WebPack build process. So, how would you suggest to go forward to have these available as HTML after build?

**Comment by EugeneHlushko** [comment]
I think what you did works well, its just the matter of **when** we generate the `_*_all.md` files. I suppose we will trigger it at some point before building html pages, but it will be trivial change since you already have made a script for it.

As we have changed the build setup in `rebuild` branch i'd like `rebuild` to get into `master` first before we introduce this change. There is last outstanding issue left and its almost done by @montogeek so rebuild should be merged into master very soon.

Meanwhile i suggest to remove the generated `_*_all.md` files from this branch and also remove the hooks from package.json as they are going to change when rebuild comes in and it will be easier to update your work with latest master at that future point.

**Comment by SetTrend** [comment]
I now removed the `*_all.md` files. I also catched-up with current master. Unfortunately something went wrong and all the `master` commits are listed now here. I don't know what went wrong and what I can do to revert my history to my original two commits.

I started a discussion at [StackOverflow](https://stackoverflow.com/questions/54174319/how-can-i-keep-my-git-history-tidy-even-after-rebasing-syncing-in-visual-studio). If I won't get a remedy there, I'm going to delete my branch and re-create it. Do you perhaps have an idea what I can do to get my branch to only show the two commits originating from myself?

File: package.json [hunk 1]
```diff
@@ -26,18 +26,19 @@
   "scripts": {
     "start": "run-s -n init:generated fetch start-only",
     "start-only": "node ./antwar.bootstrap.js develop",
-    "build": "npm run init:generated && npm run fetch && rm -rf build/ && node ./antwar.bootstrap.js build && npm run sitemap && echo webpack.js.org > build/CNAME",
+    "build": "npm run init:generated && npm run fetch && rm -rf build/ && node ./antwar.bootstrap.js build && npm run sitemap && npm run chapterize && echo webpack.js.org > build/CNAME",
```
Comment (EugeneHlushko) [inline package.json]: For now i suggest to remove this change from here and add it after rebuild is merged for less conflicts

Comment (SetTrend) [reply package.json]: Done.

File: package.json [hunk 2]
```diff
@@ -26,18 +26,19 @@
...
     "build-test": "npm run build && http-server build/",
     "deploy": "gh-pages -d build",
     "fetch": "sh src/scripts/fetch.sh",
     "init:generated": "mkdirp ./generated/loaders && mkdirp ./generated/plugins ",
     "lint": "run-s lint:*",
     "lint:js": "eslint . --ext .js,.jsx,.md",
-    "lint:markdown": "markdownlint --rules markdownlint-rule-emphasis-style --config ./.markdownlint.json *.md ./src/content/*.md ./src/content/**/*.md",
+    "lint:markdown": "markdownlint --rules markdownlint-rule-emphasis-style --config ./.markdownlint.json --ignore ./src/content/**/*_all.md *.md ./src/content/**/*.md",
```
[the remark this answers is not in the document]
Comment (SetTrend) [reply package.json]: Done.

**Comment by SetTrend** [comment]
@EugeneHlushko:

Today, I edited the `package.json` file as requested.

I'm very sorry for the long delay. I'm doing this in my spare time and was busy with other projects. Plus, I needed to learn more about git in order to accomplish the clean-up.

Cheers,
Axel

/refs:
- [How can I create merge requests after having updated my master branch?](https://stackoverflow.com/questions/54063490/how-can-i-create-merge-requests-after-having-updated-my-master-branch)
- [How can I inject an intermediate branch?](https://stackoverflow.com/questions/54139687/how-can-i-inject-an-intermediate-branch)
- [How can I keep my Git history tidy even after rebasing/syncing in Visual Studio Code?](https://stackoverflow.com/questions/54174319/how-can-i-keep-my-git-history-tidy-even-after-rebasing-syncing-in-visual-studio)
- [Add "catch-up" sync option to Git extension](https://github.com/Microsoft/vscode/issues/66462)

**Comment by EugeneHlushko** [comment]
For some reason this branch holds additional commit diff with master, lets have a look at it when we have `rebuild` merged into `master`. The worst case scenario we would cherry pick your stuff here and rebase on the updated master, therefore your contributions will be safe :)

**Comment by EugeneHlushko** [comment]
Hi @SetTrend 

I'm happy to confirm that `rebuild` finally was merged and we can proceed. Can you please update this branch and resolve conflicts? Or if its easier branch from master again and copy paste over your changes (if its quicker).

**Comment by SetTrend** [comment]
@EugeneHlushko:

I have now created a whole new fork and, after pulling the latest **webpack.js.org** `master`, I'm trying to apply my changes to the new version.

From what I can see, the HTML structure of the docs have changed. I need to apply `@media print { display: none !important; }` to these elements/classes:

```css
header,
nav,
footer,
.notification-bar,
.page-links,
.gitter
{
  @media print {
    display: none !important;
  }}
```
(... plus a new one for the Contributors section).

From what I can see, [server.jsx](/webpack/webpack.js.org/blob/master/src/server.jsx) is the page root file. Where would you want me to put SCSS for that file (i.e. `header`, `nav`, `footer`)?

**Comment by EugeneHlushko** [comment]
I would suggest `Site.scss` under `./src/components/Site/Site.scss` which imports footer, nav and header elements

**Pull request closed without merging**